### ISSUE TITLE
A problem about loss weighting and focal loss

### ISSUE THREAD

**Issue by JoJoliking** [ultralytics/yolov5#1612]
## ❔Question
Dear Author:
  I want to use yolo5 for vehicle re-identification.
  The original detector generates bbox, obj and cls for each target. I've now added a embedding vector to each target for re-identification.   
  However, when calculating the loss function, I use: hpy['cls'] * = nc / 1 （Only Cars sholud be detect. ）（In CoCo is hpy['cls'] *= nc / 80.）
  Then the loss is weighted：lembed *= h['cls'] * s
  However, such training results will lead to incorrect identification of vehicles.
  At the same time, there are a lot of high score error bounding boxes in the results.(I have tried to modify non_max_supression, but I have got the same result)
  Therefore, I would like to ask you how to choose the weight of the loss function. In other words, what effect does this weight have on the results

## Additional context
Besides, I have used focal loss and found that focal loss causes the object to be undetectable. However, the loss with BCE is normal. I don't know if I used it improperly. Can you give me some adivce?

**Comment by glenn-jocher** [comment]
@JoJoliking do not modify hpy['cls'].

**Comment by JoJoliking** [comment]
@glenn-jocher OK。But how do I calculate the weight of embedding? Re ID is actually a classification task and I only recognize the vehicle category

**Comment by glenn-jocher** [comment]
@JoJoliking I'm not sure exactly what you are trying to do, but a couple points are:

- do not modify hyp['cls']
- for single-class training there is no classification task, and hence no classification loss

I would train normally with default settings first to verify a performance baseline on your dataset before you begin making any alterations to the code.

**Comment by glenn-jocher** [comment]
@JoJoliking also in general, you should seek to balance your loss components, no matter what they are.

**Comment by JoJoliking** [comment]
@glenn-jocher Dear glenn
You mean that do not change ‘hpy['cls'] *= nc / 80’ to hpy['cls'] *= nc / 1, right（）? 
In fact, my new task of re-identifying branches is also a categorization task, so I want to know how to choose weights in YOLO5.
Last, I have used focal loss and found that focal loss causes the object to be undetectable. However, the loss with BCE is normal. I don't know if it is a bug.

**Comment by glenn-jocher** [comment]
@JoJoliking I'm really not sure what you're asking, but simply start from the basics and train your model normally to acheive a baseline as I explained above, then start experimenting.

**Comment by JoJoliking** [comment]
OK. I will have a try. Thank you for your answer

**Comment by JoJoliking** [comment]
@glenn-jocher 
Dear author
In yolov5, the learning rate is calculated and decreased automatically. If I need to use the fixed Lr0 to train in the next 10 epoch after warmup and then use your method to calculate the learning rate, how should I modify it.
Thank you for your kind reply every time. Looking forward to your further work!

**Comment by glenn-jocher** [comment]
@JoJoliking you can control the LR scheduler here. The warmup code is programmed to slowly meet this LR whereever it is after the warmup period:

https://github.com/ultralytics/yolov5/blob/94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5/train.py#L126-L131

**Comment by JoJoliking** [comment]
@glenn-jocher 
Can you be more specific?I have tried to understand the changing principle of learning rate, but I still can't understand it.Since I use hyp.finetune.yaml, I want the learning rate to keep lr0=0.0032 in the 10 epoch after warmup.
At last. Thank you very much for your help and guidance.I have learned a lot about YOLO at present

**Comment by glenn-jocher** [comment]
@JoJoliking no I can't be more specific. Modifications are up to you, make any you want and then use the provided plotting function to visualize the resutls. LRs are also displayed in the loggers (Tensorboard and W&B).

**Comment by JoJoliking** [comment]
OK. I will have a try again
